### PR TITLE
job manager: add job states, exceptions

### PR DESCRIPTION
**Pull request by garlick**
This is some cleanup to the job manager, the `flux-job` command, and the libjob API, driven in part by the state and exception descriptions in [RFC 21](https://github.com/flux-framework/rfc/blob/master/spec_21.adoc)

There is now a `flux job raise` command that can raise arbitrary exceptions:
```
Usage: flux-job raise [OPTIONS] id [message ...]
Raise exception for job
  -h, --help             Display this message.
  -s, --severity=N       Set exception severity [0-7] (default=0)
  -t, --type=TYPE        Set exception type (default=cancel)
```
`flux job cancel` is now an alias to `flux job raise`.  Only exceptions with `severity=0` have an effect on the job at this point.

The `flux_job_cancel()` API function is replaced with:
```c
/* Raise an exception for job.
 * Severity is 0-7, with severity=0 causing the job to abort.
 * Note may be NULL or a human readable message.
 */
flux_future_t *flux_job_raise (flux_t *h, flux_jobid_t id,
                               const char *type, int severity,
                               const char *note);
```

Jobs now have states, now reported with a column in `flux job list` output:
```
$ flux job list
JOBID		STATE	USERID	PRI	T_SUBMIT
96552878080	C	5588	16	2019-02-12T22:43:45Z
165490458624	C	5588	16	2019-02-12T22:43:49Z
3035736571904	N	5588	16	2019-02-12T22:46:40Z
```
At the moment jobs are created in `NEW` state, and they transition to `CLEANUP` when you cancel them.

The "job state flags" are gone from the public API and from the `flux job list` output.

There was some additional cleanup, mainly in the job manager.

I had these changes mixed in with the scheduler interface stuff, but thought it best to submit as its own PR.

I'll call it a WIP at this point because I think I can improve the testing a bit.

### CODE REVIEW

**Comment by grondo** [comment]
This is nice! :+1:

**Comment by garlick** [comment]
Just forced pushed a rebase on current master, the addition of some testing (mostly unit tests), and a couple of minor bug fixes uncovered in testing.

**Comment by garlick** [comment]
I'm not sure what's going on with travis.  Looks like the tests have all passed but the little ball is still orange.  Hrm.

**Comment by grondo** [comment]
I checked on travis and it says the status of your pr is "snowed in"

**Comment by garlick** [comment]
That seems to have done the trick.

File: src/cmd/flux-job.c [hunk 1]
```diff
@@ -104,11 +108,18 @@ static struct optparse_subcommand subcommands[] = {
       NULL,
     },
     { "cancel",
-      "[OPTIONS] id ...",
-      "Abort job(s)",
-      cmd_cancel,
+      "[OPTIONS] id [message ...]",
+      "Alias for flux job raise",
+      cmd_raise,
```
Comment (grondo) [inline src/cmd/flux-job.c]: I think it is ok to hid the fact that `flux job cancel` is an alias for `flux job raise` and provide help message that says "Cancel a job" for better UI.

I love that you can provide a message when you cancel a job.

File: src/modules/job-manager/util.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * Copyright 2018 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* active - manipulate active jobs in the KVS
```
Comment (grondo) [inline src/modules/job-manager/util.c]: minor nit: I assume this line is left over from when this file was `active.c`, should the comment read `util - manipulate active jobs in kvs` or similar?

File: src/common/libjob/job.h [hunk 3]
```diff
@@ -74,9 +71,13 @@ int flux_job_submit_get_id (flux_future_t *f, flux_jobid_t *id);
...
 
-/* Abort a job.
+/* Raise an exception for job.
+ * Severity is 0-7, with severity=0 causing the job to abort.
+ * Note may be NULL or a human readable message.
  */
-flux_future_t *flux_job_cancel (flux_t *h, flux_jobid_t id, int flags);
+flux_future_t *flux_job_raise (flux_t *h, flux_jobid_t id,
+                               const char *type, int severity,
+                               const char *note);
```
Comment (grondo) [inline src/common/libjob/job.h]: Any reason not to keep a `flux_job_cancel` as an user-friendly alias for `flux_job_raise`? (I imagine an exception of type `cancel` will be over 90% of use cases, and saving users from writing `flux_job_raise (h, id, "cancel", 0, NULL)` will be a mercy)

Could even just be a macro at this point.

File: src/cmd/flux-job.c (hunk 1 above)
Comment (garlick) [reply src/cmd/flux-job.c]: Sounds good.

File: src/modules/job-manager/util.c (hunk 2 above)
Comment (garlick) [reply src/modules/job-manager/util.c]: Oops! ! Yes sloppy error.

File: src/common/libjob/job.h (hunk 3 above)
Comment (garlick) [reply src/common/libjob/job.h]: OK, will do.

**Comment by garlick** [comment]
Rebased on current master, fixed review comments, and pushed a couple of other minor little things that should be squashed down before this is merged.

**Comment by grondo** [comment]
Looks like a builder hit #2015

**Comment by garlick** [comment]
Great thanks!

On `flux job list` - JSON output is a good idea as it would let us write tests that are not dependent on column position of output.    I'll open an issue on that one.

**Comment by garlick** [comment]
Rebased after @chu11's fix, and also squashed down some incremental development.

**Comment by garlick** [comment]
Rebased.  (Thanks guys for the nice feedback!)

**Comment by grondo** [comment]
Just needs a rebase and I think it is good to go in!

**Comment by garlick** [comment]
Thanks, rebased and also snuck in a fix for this typo (squashed):

```diff
diff --git a/src/modules/job-manager/raise.c b/src/modules/job-manager/raise.c
index 861d4a5e..ffa6cea7 100644
--- a/src/modules/job-manager/raise.c
+++ b/src/modules/job-manager/raise.c
@@ -79,7 +79,7 @@ void raise_ctx_decref (struct raise_ctx *c)
 {
     if (c && --c->refcount == 0) {
         int saved_errno = errno;
-        c->job->flags &= JOB_EXCEPTION_PENDING;
+        c->job->flags &= ~JOB_EXCEPTION_PENDING;
         if (c->severity == 0)
             c->job->state = FLUX_JOB_CLEANUP;
         if (c->request) {
```

**Comment by garlick** [comment]
Yesss!   Thanks.

**Pull request merged**